GH-3945: Fix `not eligible for getting processed`

artembilan opened the pull request:
Fixes https://github.com/spring-projects/spring-integration/issues/3945

The `IntegrationManagementConfiguration` produces an `IntegrationManagementConfigurer` which is a `BeanPostProcessor`. According to Spring recommendation this kind of infrastructure beans must be declared as `static`. Due to an `implements ImportAware, EnvironmentAware` nature of the `IntegrationManagementConfiguration`, we cannot use `static @Bean` method.
But since the `IntegrationManagementConfiguration` is not involved in any bean post-processing, it is safe to follow recommendation and mark it as a `@Role(BeanDefinition.ROLE_INFRASTRUCTURE)`.

* Fix `MessagePublishingInterceptor` to initialize `MessagingTemplate` and `DestinationResolver` lazily
* Fix `AbstractMethodAnnotationPostProcessor` to initialize `DestinationResolver` lazily

**Cherry-pick to `5.5.x`**

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-3945: Fix `not eligible for getting processed`

Fixes https://github.com/spring-projects/spring-integration/issues/3945

The `IntegrationManagementConfiguration` produces an `IntegrationManagementConfigurer` which is a `BeanPostProcessor`.
According to Spring recommendation this kind of infrastructure beans must be declared as `static`.
Due to an `implements ImportAware, EnvironmentAware` nature of the `IntegrationManagementConfiguration`,
we cannot use `static @Bean` method.
But since the `IntegrationManagementConfiguration` is not involved in any bean post-processing,
it is safe to follow recommendation and mark it as a `@Role(BeanDefinition.ROLE_INFRASTRUCTURE)`.

* Fix `MessagePublishingInterceptor` to initialize `MessagingTemplate` and `DestinationResolver` lazily
* Fix `AbstractMethodAnnotationPostProcessor` to initialize `DestinationResolver` lazily

**Cherry-pick to `5.5.x`**

garyrussell reviewed (changes requested):
Don't we need test(s) to verify that the behavior being fixed no longer occurs?

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/aop/MessagePublishingInterceptor.java:
```diff
@@ -143,6 +143,15 @@ public final Object invoke(final MethodInvocation invocation) throws Throwable {
 	}
 
+	private void initMessagingTemplateIfAny() {
+		if (this.templateInitialized.compareAndSet(false, true)) {
```
Ah! I see. 
Yeah... I questioned myself why in other place we use `boolean` flag instead of this `AtomicBoolean`.
Thank you! 😄

artembilan committed:
* Use `getChannelResolver()` internally in the `AbstractMethodAnnotationPostProcessor`
instead of direct property access which might not be initialized yet
* Use a plain `boolean` for `templateInitialized` in the `MessagePublishingInterceptor`
to avoid skips in other thread where and move on with still not initialized properties

artembilan committed:
* Remove unused import

artembilan committed:
* Fix `this.` prefix for `beanFactory` property reference

garyrussell commented:
Cherry-pick is not clean - please back port.

artembilan commented:
... and cherry-picked to `5.5.x` as https://github.com/spring-projects/spring-integration/commit/067de96ab34af245994786241979b13e6e16e13c after fixing conflicts.
